### PR TITLE
chore: make request queue types more accurate

### PR DESCRIPTION
**Pull request by vladfrangu**
CC: @pocesar @B4nan @mnmkng

### CODE REVIEW

File: src/resource_clients/request_queue.ts [hunk 1]
```diff
@@ -94,11 +94,11 @@ export class RequestQueueClient {
         return this.emulator;
     }
 
-    async get(): Promise<Record<string, unknown> | undefined> {
+    async get(): Promise<RawQueueTableData | undefined> {
```
Comment (mnmkng) [inline src/resource_clients/request_queue.ts:97]: Since this is a `requestQueueClient.get()` function, I think it would make sense if it returned `RequestQueue`.

The thing is, there's a mirror of this function in the `apify-client` library. And both `apify-client` and `@apify/storage-local` are used in the Apify SDK interchangeably.

So, maybe @B4nan could help me here, is there gonna be a problem if one of the libraries will return a different type than the other? Do we need to export some shared types somewhere?

Finally a small nitpick, can you write `Promise<?Something>` instead of `Promise<Someting | undefined>` it TypeScript? The `| undefined` looks so lame 😄

File: src/emulators/request_queue_emulator.ts [hunk 2]
```diff
@@ -150,14 +172,14 @@ export class RequestQueueEmulator {
...
-    selectModifiedAtById(id: string | number): unknown {
+    selectModifiedAtById(id: string | number): Date {
         if (!this._selectModifiedAtById) {
             this._selectModifiedAtById = this.db.prepare(`
                 SELECT modifiedAt FROM ${this.queueTableName}
                 WHERE id = ?
             `).pluck();
         }
-        return this._selectModifiedAtById.get(id);
+        return new Date(this._selectModifiedAtById.get(id));
```
Comment (mnmkng) [inline src/emulators/request_queue_emulator.ts:182]: Why the change of types here? Can't it return `string` like it always did?

Comment (vladfrangu) [reply src/emulators/request_queue_emulator.ts:182]: The type for this function has been marked as `Date` before the TS conversion, then I found out the SQL data was a string (which wasn't compatible with the types), so now it's technically correct

File: src/resource_clients/request_queue.ts (hunk 1 above)
Comment (B4nan) [reply src/resource_clients/request_queue.ts:97]: > So, maybe @B4nan could help me here, is there gonna be a problem if one of the libraries will return a different type than the other? Do we need to export some shared types somewhere?

TS is structurally typed, so its enough to actually share the same interface without denoting it. See this example:

https://www.typescriptlang.org/play?#code/PTAEHcEsBcAtUgO2gUwE4DMCGBjFBnAKCVU1xVAEFQBvQ0B0DAe2YC5R9o0kBzAbnqMARljQdEAVwC2w9IIC+hYsnTY8oAEK0hDFu07c+gxqFHjQU2fMJLCAGxTRQWDtQC8tJqw4ByLL4ANGZiHACMoAqCjs7CHNqeNN4GvsJBIRYATJGCWKCewoKEIBDMaADW+BAw8Dj2WPj4BKDQrIR1DVWU1HSm+gCEHFw8iAK6GYOWMnJoisodjVravYwDQ0ajJiJik1Yzcw5OLq5UHl76fgHB5uE5h7FxS-nnPqCp6Teg2VGEWHkFhSAA

> Finally a small nitpick, can you write Promise<?Something> instead of Promise<Someting | undefined> it TypeScript? The | undefined looks so lame 😄

The TS way would be to have a generic type like `Maybe<T> = T | undefined | null` and use `Promise<Maybe<Something>>`. I am not aware of better solution for return types (for arguments there is `?` so you can do `prop?: string`).

Comment (vladfrangu) [reply src/resource_clients/request_queue.ts:97]: > Finally a small nitpick, can you write `Promise<?Something>` instead of `Promise<Someting | undefined>` it TypeScript? The `| undefined` looks so lame 😄

No, you cannot
image

File: src/emulators/request_queue_emulator.ts (hunk 2 above)
Comment (mnmkng) [reply src/emulators/request_queue_emulator.ts:182]: Oh, I see. The type was wrong then. I guess for consistency with the other functions, which always return whatever the database returns, we should have `selectModifiedAtById(id: string | number): string {` and if the consumer function needs a `Date`, we should do the conversion there.

File: src/resource_clients/request_queue.ts (hunk 1 above)
Comment (mnmkng) [reply src/resource_clients/request_queue.ts:97]: Ok, if there's no standard way then let's keep it as it is.

File: src/emulators/request_queue_emulator.ts (hunk 2 above)
Comment (mnmkng) [reply src/emulators/request_queue_emulator.ts:182]: Yes please 👍👍👍 , have the `selectModifiedAtById`  function return `string` and then change it to `Date` in `listHead` where it's used.
